### ISSUE TITLE
Link Control: Add more options.

### ISSUE THREAD

**Issue by paaljoachim** [WordPress/gutenberg#23011]
**Is your feature request related to a problem? Please describe.**

Many have been asking for an easier way to have additional link options.

Add easy way to modify links and add toggle options.
#17556

Add nofollow option for links
#13542

Make link selection consistent
#17293

and there are probably other associated issues.

**Describe the solution you'd like**

Here is one way to do so. Adding a more icon to bring in additional link options.

Added icon.
![Screen Shot 2020-06-08 at 23 57 45](https://user-images.githubusercontent.com/5323259/84084526-f3224080-a9e3-11ea-8402-c8d04517207a.png)

Hover icon.
![Screen Shot 2020-06-08 at 23 57 57](https://user-images.githubusercontent.com/5323259/84084536-f9182180-a9e3-11ea-8d0c-20037665f291.png)

Opening additional options.
![Screen Shot 2020-06-08 at 23 58 09](https://user-images.githubusercontent.com/5323259/84084587-09c89780-a9e4-11ea-80cd-6d9845813b03.png)

I am not sure which options should be added to the more screen, but I believe it would be helpful for many that it does exist. Perhaps only adding a way for plugins to add the more option or having it directly added to Gutenberg Core.

**Comment by arnaudbroes** [comment]
We did this with AIOSEO (both toggle controls and a text field) but honestly it was a real pain to do. It would be great if the Gutenberg team added any way for plugin developers to extend the existing options with their own.

**Comment by paaljoachim** [comment]
I'll also let Dave and Marcus know about this issue...
@getdave and @mkaz 
As I do think it can solve a few issues in regarding link controls.

**Comment by karmatosed** [comment]
I think it would be important to think what the reason for adding more link options is and if they are even needed. I don't believe by default more options are helpful myself. I also don't think hiding if they are useful makes sense. In this case I would suggest examining what is needed over creating another hidden area to expose options. My recommendation would be to start from the point of why is something needed?

**Comment by phpbits** [comment]
> I think it would be important to think what the reason for adding more link options is and if they are even needed. I don't believe by default more options are helpful myself. I also don't think hiding if they are useful makes sense. In this case I would suggest examining what is needed over creating another hidden area to expose options. My recommendation would be to start from the point of why is something needed?

Agreed! Classic editor doesn't have these options and always been plugin territory. We've discussed this before and as I recalled we agreed on having filter/hook on the link control for developers to extend; and be able to add these features easily.

**Comment by phpbits** [comment]
> We did this with AIOSEO (both toggle controls and a text field) but honestly it was a real pain to do. It would be great if the Gutenberg team added any way for plugin developers to extend the existing options with their own.

💯. To be able to make this work there's a need to create separate format and rely on custom URL popover. 

@arnaudbroes Looks like you are using the one I've created for EditorsKit plugin. Would really appreciate if you can support EditorsKit ;) Thanks!

**Comment by richtabor** [comment]
Think of this more of a way for third-party plugins to add link controls here, such as a "No follow" toggle, or "Sponsored" toggle (or whatever a plugin may want to add). I'd say that if there were no additional controls added to the filter, nothing changes from how the link controls look today. 

BUT, if there _are_ additional controls added to a future filter, then the "More link controls" arrow could add those additional controls. 

Sorta the best of both worlds by letting folks essentially configure their experience. I believe adding more top-level controls here would just further complicate the interface - especially if those controls are no where near the primary action (no follow for example).

**Comment by getdave** [comment]
Firstly thanks for raising this @paaljoachim. Much appreciated.

> I recalled we agreed on having filter/hook on the link control for developers to extend; and be able to add these features easily.

When I originally worked on this component this was the approach we were going to take. However, I seem to recall we (I) ran out of "space" to accommodate in the PR. Apologies for this.

Could we not try the following:

1. Add the hook to afford extensibility. No harm and solves many issues.
2. Do some user research (not sure how) to see if there are any extremely common controls that would be appropriate to add as _defaults_. Who would do this and how are beyond my scope of expertise and knowledge.

For me, the issue is that we can't service _every_ requirement in Core without ending up with a completely cluttered link UI. That is something we should avoid. Plugins are a good fit for many of these options IMHO.

**Comment by richtabor** [comment]
Here's an [updated Figma Prototype](https://www.figma.com/proto/11zYmEHtoQ7mejqwzdLsIS/More-Link-Controls?node-id=1%3A7&viewport=582%2C647%2C1&scaling=scale-down) and modified mockup.

<img width="872" alt="Screen Shot 2020-06-09 at 11 47 00 AM" src="https://user-images.githubusercontent.com/1813435/84169742-fa2e6a80-aa46-11ea-89b9-86ad8f1ab0b0.png">

**Comment by arnaudbroes** [comment]
> @arnaudbroes Looks like you are using the one I've created for EditorsKit plugin. Would really appreciate if you can support EditorsKit ;) Thanks!

That's correct, it's inspired on your great work, although we made quite a bit of changes. We also added support for the new component that was included in WP v3.4.0. Feel free to have a look at it!

**Comment by arnaudbroes** [comment]
>  I don't believe by default more options are helpful myself.

A"nofollow" option is quite useful though. I might be mistaken but I believe an earlier iteration of the previous default editor in WordPress had an option to add this. That's why I'm personally not fully convinced that this should be plugin territory.

> I also don't think hiding if they are useful makes sense. In this case I would suggest examining what is needed over creating another hidden area to expose options. 

Fully agree on this. If they're added, they shouldn't be hidden. I do understand that we want to strike a balance between extra options and UI clutter.

**Comment by arnaudbroes** [comment]
> Here's an [updated Figma Prototype](https://www.figma.com/proto/11zYmEHtoQ7mejqwzdLsIS/More-Link-Controls?node-id=1%3A7&viewport=582%2C647%2C1&scaling=scale-down) and modified mockup.

What concerns me here is that this might be very intuitive for the average user if they're hidden by default, although you could argue these options are meant for more "advanced" editors.

**Comment by richtabor** [comment]
> What concerns me here is that this might not* be very intuitive for the average user if they're hidden by default, although you could argue these options are meant for more "advanced" editors.

I'm not sure the everyday WordPress user would understand what a no-follow link does, though this would be nice to test for discoverability. 

I'm just so hesitant on continuously adding to the current UI – its already difficult to navigate.

**Comment by jdevalk** [comment]
Relates to #16609 as well, which was closed, but which caused both AIOSEO, and Editor Kit, and us at Yoast, to build something which is really really ugly and not what we wanted to do. I'm all for calling this plugin territory as I don't think we'd implement it the same way even between those plugins, but it should be _possible_ to do at least. So, mostly, I want this to be filterable / slot fillable.

**Comment by karmatosed** [comment]
Looping back to this, am I right that the path forward would be to have it as hooks and then consider from there if core has an interface for it (or not)? That said, this should change to hooks specifically and have a dev flag I feel, so going to do that and move it on from design feedback as the interface isn't something to be considered right now.

**Comment by maartenhunink** [comment]
Being able to add additional attributes to a link would be very useful. For some of the websites I work on this would enable me to create more advanced links. I assume this feature request is waiting for somebody to pick it up?

**Comment by paaljoachim** [comment]
Hey @shaunandrews I am going to loop you into this issue as well. As you are exploring the link control UI.
https://github.com/WordPress/gutenberg/pull/28755

**Comment by TeamDNK** [comment]
We also need the ability to add custom attributes to the <a> element. Toggle buttons may too limited for cases that require more than a boolean return. The block filters seem to be a great approach to extend the popover so that developers are able to add elements to it or filter the entire popover.

**Comment by altivero** [comment]
Add media link as per the following [mock-up](https://github.com/WordPress/gutenberg/issues/8322#issuecomment-464841591) in #8322. Perhaps even an option to search for media already in the library.

**Comment by paaljoachim** [comment]
There has been a lot of discussion in this closed issue lately:
Insert Link: consider adding an "Additional CSS Class" field to the link inserter
https://github.com/WordPress/gutenberg/issues/13368

It sounds like we also need to add a Link CSS Class similar to what is done for the link control settings with images.

**Comment by paaljoachim** [comment]
Here is a modified link control settings initially created by @richtabor 
Where I also added the Link CSS Class field.


![Screenshot 2022-02-22 at 13 49 27](https://user-images.githubusercontent.com/5323259/155135722-681f7292-4a64-42a4-9223-8f86ee33bfdf.jpg)

![Screenshot 2022-02-22 at 13 49 16](https://user-images.githubusercontent.com/5323259/155135747-2edbf750-61b3-48c2-90c8-48d1f6007499.jpg)

**Comment by wpmarmite** [comment]
In Elementor, the link component looks like this:

<img width="282" alt="Capture d’écran 2022-09-20 à 15 42 05" src="https://user-images.githubusercontent.com/15243992/191274166-77a44d45-13af-470e-bfc2-97796515df53.png">

I don't believe that core needs to go this far as it could disturb everyday users.

However, plugin developers should be able to enhance this component (with a dedicated filter) to add link options to their users (`nofollow`, `sponsored`, add a custom `class`, `id`, or even an `aria-label` attribute).

**Comment by erik-visser** [comment]
Very interesting read. And I get the dilemma. I would like to add an additional argument for it that I didn't see in the comments. It's also good for IndieWeb purposes and using web standards for contextualizing links. And I think it should be in the interest of Wordpress to give a user the option to use those standards. If it should be part of the core, I don't know but adding yet another plugin to the growing collection isn't always the best option either. Thanks for all your great work people!

**Comment by frogdesk** [comment]
NEED classes!!!

https://github.com/WordPress/gutenberg/issues/38878
https://github.com/WordPress/gutenberg/issues/13368

**Comment by getdave** [comment]
My view is that Gutenberg should offer a sensible set of common default settings. Would this suffice?

- Open in new tab
- No follow
- CSS classes

**Comment by richtabor** [comment]
> CSS classes

I'm not certain on this, but I would like more context of the use-case of adding support for this.

**Comment by pfefferle** [comment]
I think besides the styling possibilities (using CSS classes), the rel and class attribute is important to add Microformats!

See:

* https://gmpg.org/xfn/ (rel-attribute)
* https://microformats.org/wiki/h-entry#Proposed_Additions

Which is important for the IndieWeb movement: https://indieweb.org/

**Comment by T4ngml** [comment]
It's now been YEARS I pray for my redactors to be able to apply specific designs for internal and external links.

Let's get back to basics.

If we folllow the rules, only external links should, potentially, be open in new tabs.
And the very vast majority of links we might want to set as no follow is external links. Not to say we'd like to make them all no follow, but that's off topic...

Beyond that, there are some cases, where, **advanced content editors** sometime need to step away from these rules:
- For SEO purposes, I sometime want to add many links in a content. But finaly realize it's actually a bit too much for the reader, so sometimes want some of them not noticable
- For SEO purposes, I can also, sometimes, want to play with the no follow and new tab behaviors.

I feel like, a non-techy option, could basically be to add some logic, which applies specific classes to both internal, and external links. So that some global CSS rules could make internal and external links visually differentiable without the user's intervention... Or remain not noticable when no specific CSS styling rules are set for this.

This could be provided via:
- an "advanced options" toggle in the Inser-/edit link pop-in
- an option in WordPress settings to make them available in the pop-in
- just leave it to plugins.

What's certain though, is that in the meatime:
- non-HTML skillful users aren't able to provide a consistent styling for their links. The most zealous among them will manually apply manually, repeatedly, religiously, a set of many visual transformations, with neither hover nor responsive capabilities. 
- Advanced users keep yelling at Gutenberg everytime it destroys the tiny perfect job they meticulously adjusted in the HTML editor. Then try to step back with Ctrl+Z, or versions, then yell even more
- Medium skilled developers keep delivering and maintain monkey patches, to style the content the way the user want it... with unsatifsfying results.

**Comment by frogdesk** [comment]
I just can't even fathom why you think it was a good idea to remove the ability to add classes to links in the first place.

This is used for all kinds of things, including modals, plugins that add roles or other attributes based on classes, and simply for styling purposes. Links NEED to be able to have classes, there are several issues FULL of people begging for this functionality to be returned.

**Comment by Pairfum-London** [comment]
This is how Yoast implemented this:
![image](https://user-images.githubusercontent.com/60850234/218081411-0110c3c9-3500-487d-a379-31835fbb90ad.png)

From a user's perspective this is easy to use.

**Comment by jeremyfelt** [comment]
Answering @richtabor's prompt specifically re: additional CSS classes

> I'm not certain on this, but I would like more context of the use-case of adding support for this.

First, I agree with @pfefferle's IndieWeb point. The most immediate (most easily described? 🤔) way for people to start getting familiar with IndieWeb is by adding [microformat](https://microformats.org/wiki/microformats2) class names directly to elements.

The block editor generally makes this a nice experience because you can assign classes like `h-card`, `e-content`, etc... to almost any container without waiting for a theme developer to fully implement microformats.

To properly markup [replies](https://indieweb.org/in-reply-to), [likes](https://indieweb.org/like), [bookmarks](https://indieweb.org/bookmark), etc... a person must be able to add `u-in-reply-to`, `u-like-of`, `u-bookmark-of` to the anchors linking to others' content. This is not currently possible, or a very unstable experience when switching between code/visual.

Second—_and actually the real reason I'm here even though I think the IndieWeb example is more compelling_...

I'm working on integrating a plugin written by a SaaS company that implements a membership login flow with their JavaScript. That JavaScript looks for specific class names on anchors.

I've created the markup for this as a template part so that the site owner can update the text surrounding those anchors, but when I open it in the template editor, I get a block validation error because there are class names on the anchor that the template editor can't handle.

I'll likely need to create a couple custom blocks for server rendered content so that I can be confident those class names will remain when this content is edited. This is probably more reliable than shimming things with JavaScript so that class names are added to those anchors before the 3rd party script runs. Even though I'm happy I have an approach to solve the problem, both of those options seem like an unnecessary amount of work for ensuring a class name is on an anchor.

Probably related: #10235, which _could_ make it easier to implement all of this as a custom inline block.

**Comment by frogdesk** [comment]
We will be switching away from WordPress for future clients because of this.

**Comment by getdave** [comment]
As I have said previously, I believe we should look to add this feature. Once in the Plugin it can always been rejected if enough people feel strongly that it shouldn't be included.

Please bear in mind that @richtabor is entirely right to question the use case for adding new features to the editor. Doing this is an important way to ensure only features that are truly needed are included to avoid a bloated UI.

Nonetheless I appreciate the strength of feeling expressed by commenters here as well as the clear use cases.

If there isn't already let's open a dedicated Issue for adding CSS classes to the Link Control. This one is too broad.

**Comment by ghost** [comment]
Is any dev actively working on this, and if not, what needs to be done? It shouldn't be that hard to add a nofollow toggle here right? I can try to add it tomorrow and submit a PR. Would at least fix part of this issue.

<img width="406" alt="Screenshot 2023-03-02 at 11 37 39 PM" src="https://user-images.githubusercontent.com/118618474/222575762-c21ea385-fa94-4b2e-999c-9baf0a7df6b2.png">

**Comment by frogdesk** [comment]
> Please bear in mind that @richtabor is entirely right to question the use case for adding new features to the editor. Doing this is an important way to ensure only features that are truly needed are included to avoid a bloated UI.
> 
> Nonetheless I appreciate the strength of feeling expressed by commenters here as well as the clear use cases.

It's not right to question the use case for adding classes to links. That is something only someone with zero development experience would do.

**Comment by getdave** [comment]
> Is any dev actively working on this, and if not, what needs to be done? It shouldn't be that hard to add a nofollow toggle here right? I can try to add it tomorrow and submit a PR. Would at least fix part of this issue.

I'm not aware of anyone actively working on adding more controls at this point. However, I have been working to revise the UI/UX to prepare the ground for adding more controls as one of the main objections in the past is that it would add too much noise to the UI.

Now I've added a settings "drawer" it will be easier to add more controls such as you suggest.

I was planning on circling back to this once WP 6.2 is launched as currently it's an extremely busy period for contributors.

> It shouldn't be that hard to add a nofollow toggle here right?

It would be great to receive a PR from you. Thank you! This is however more than simply adding the toggle. You will then need to ensure that the resulting data is _persisted_ in every instance where the toggle is used.

**Comment by ghost** [comment]
> It would be great to receive a PR from you. Thank you! This is however more than simply adding the toggle. You will then need to ensure that the resulting data is persisted in every instance where the toggle is used.

@getdave True, might underestimate it, but I'll take a shot. Will let you know if it's out of my league at the time being. Thanks!

**Comment by ghost** [comment]
First PR is up now. Not complete, but we're getting closer. Let's move this discussion to the PR #48722 

https://user-images.githubusercontent.com/118618474/222710020-770b2066-8629-40ef-8f6c-a59044f1281f.mov

**Comment by richtabor** [comment]
I'm leading a proposal to clean-up the experience of editing links in https://github.com/WordPress/gutenberg/issues/50890 which will make further link settings (i.e. potential no-follow and css class functionality) much more feasible. Would appreciate a look.

**Comment by adamwoodnz** [comment]
> I'm leading a proposal to clean-up the experience of editing links in #50890 which will make further link settings (i.e. potential no-follow and css class functionality) much more feasible. Would appreciate a look.

Noting that https://github.com/WordPress/gutenberg/issues/50890 didn't address adding CSS class to links unfortunately. This remains a [requested feature in the support forums](https://github.com/WordPress/gutenberg/issues/38878#issuecomment-1678111822).

**Comment by richtabor** [comment]
> Noting that https://github.com/WordPress/gutenberg/issues/50890 didn't address adding CSS class to links unfortunately.

Correct, that UX provides an affordance for adding additional controls. :)

**Comment by frogdesk** [comment]
Is there some kind of misconception that this is a requested feature? Incorrect, this is a breaking regression that should be considered an urgent bug fix. There was a working function to add classes to links previously and it was removed, completely breaking existing workflows.

**Comment by acirujano** [comment]
I want to add a UX perspective that reinforces the need for adding a “CSS class” field to the Link UI popover.

### 1. Historical context: this used to be possible
This is not a new feature request. As noted in #13368 and discussed here in #23011, Gutenberg previously allowed adding a class to an inline link via HTML editing, and some workflows relied on this. When the behavior changed, it removed a capability that users depended on.

### 2. The current workaround (HTML mode) is fragile
The expected workaround is switching to the Code editor and manually adding the class. In practice:

- Editing HTML can lead to errors when switching back to the visual editor.  
- In some cases (see #27790), updating link attributes such as “open in new tab” causes manually added classes to be removed.  
- This breaks user expectations and creates friction for authors who are not comfortable editing raw HTML.

From a UX standpoint, this is not a reliable or predictable workflow.

### 3. Why a class field belongs in the Link UI popover

- The Link UI is already where users expect to configure link-level attributes (`rel`, `target`).  
- A CSS class is semantically attached to the `<a>` element, not to the block containing it.  
- Relying on block-level classes forces users to style *all* links in a block, which does not reflect common authoring needs.

Providing a class field directly in the popover would align with existing mental models and avoid switching contexts.

### 4. Clear real-world use cases that authors encounter every day
- Adding icons, tooltips, or visual indicators to specific links.  
- Styling individual calls to action inside paragraphs or lists.  
- Attaching accessibility enhancements or ARIA-driven behaviors to one link.  
- Tracking clicks on specific links without relying on block-level classes.  
- Implementing design tokens and microinteractions on a per-link basis.
- Attaching accessibility enhancements or ARIA-driven behaviors to one link (e.g., screen reader enhancements, keyboard navigation patterns).

These are basic, legitimate use cases that authors cannot currently fulfill without risking editor breakage.

### 5. Proposal
Add a “CSS class” field under the **Advanced** section of the Link UI popover, consistent with how block classes are exposed, and extending the existing `rel` and target controls.

This would restore an editing capability that users previously had, reduce reliance on fragile HTML workflows, and improve author experience across many common patterns.

**Comment by getdave** [comment]
Thanks for your proposal. I added what you have requested in https://github.com/WordPress/gutenberg/pull/67560 and it shipped with WordPress 7.0.

I hope that helps?

I think we can close this Issue now. Please feel free to reopen if there's still elements that are outstanding.

**Comment by pfefferle** [comment]
@getdave I think the request to add more controls is still valid and not yet solved. I appreciate your work allowing users to add classes, but I think this is not very end-user friendly.

The IndieWeb for example makes heavy use of classes to define a "reply context", like is the linking a "[like](https://indieweb.org/like)", a "[reply](https://indieweb.org/reply)" or simply a "mention". It would be nice to extend the Link Controls to add toggles or dropdowns to implement features like that.

**Comment by getdave** [comment]
I understand. However, those features are unlikely to be come something Core provides. 

I think perhaps the larger feature is making this area potentially extendable by _Plugins_.

However that seems outside the scope of this Issue itself.

**Comment by pfefferle** [comment]
@getdave after re-reading my comment, I think it was a bit misleading! My intention was to request the extensibility for other plugins to hook into the link-controls to provide some extra possibilities. The IndieWeb reference was only an example. Sorry for the confusion!

What I tried to say is: "I think the request to **be able to** add more controls is still valid and not yet solved"!

**Comment by getdave** [comment]
No problem! I agree with your proposal/request but I think this Issue is no specific enough and doesn't fully outline the requirement.

Would you be open to raising a new Issue which captures this? 

Also worth noting the root problem is not the extensibility of the `LinkControl` component per se, but rather the ability to control how the data is mapped to persistence.

For example, the inline link format which is what is used for all links in (for example) Paragraph blocks has no extensibility mechanic to allow Plugin devs to refine how the link format fields map to attributes on the block.

I started on something here https://github.com/WordPress/gutenberg/pull/54791 but I never had time to advance it.
